src/Smidge/SmidgeStartup.cs

```diff
@@ -85,27 +101,68 @@ public static class SmidgeStartup
             services.AddTransient<BundleRequestModel>();
             services.AddTransient<CompositeFileModel>();
 
+#if NETCORE3_0
+            //tell .net core 3.0 where to find our controller
+            var assembly = typeof(SmidgeStartup).Assembly;

```

Comment:
I'm unsure about this block of code and the `regControllersAsService` parameter. It seems like just manually registering the controller would suffice? `services.TryAddTransient(typeof(SmidgeController));`

Reply:
I found that if AddControllersAsServices had been called elsewhere, then you had to call it again otherwise it would not find the controllers. There is no way to detect if that had been done already (outside the call to AddSmidge). Just adding the controllers as transient services didn't work for me.. but that was back in preview5 so it's possible it will work now, worth testing.

Reply:
Ok thanks, i'll investigate some more. Have been diving into the source code for some of this and starting to make some sense out of things. Just to be sure though, the reason calls would be made to AddControllersAsServices is if you opt in to use `ServiceBasedControllerActivator` instead of the default one?

Reply:
Have had a look through all the source code for this stuff and doing what is here seems fine, though IMO we should remove the `regControllersAsService` and just ensure that people call this after they use `AddSmidge`. 

I was hoping it would be more simple to add our own 'parts' to the app and IMO it should be since what happens under the covers when we do this will work but it's also allocating a bunch more object instances and it's also re-checking a huge amount of stuff exists within the container. 

Do `AddController` does [this](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs#L87) and calls [AddControllersCore](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs#L141) which calls `AddMvcCore`, `AddApiExplorer`, `AddAuthorization`, `AddCors`, `AddDataAnnotations`, `AddFormatterMappings` and each of those methods goes on to do a huge amount of checks. 

At this stage, all we need is an instance of `IMvcCoreBuilder` and there's no simple way to get that without re-writing this method: [`GetApplicationPartManager`](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L74) which is a bit ugly. 

So the code we can use that will have a minimal impact is this:

```cs
var builder = services.AddMvcCore();
builder.ConfigureApplicationPartManager(manager =>
{
    var assembly = typeof(SmidgeStartup).Assembly;
    var partFactory = ApplicationPartFactory.GetApplicationPartFactory(assembly);
    foreach (var applicationPart in partFactory.GetApplicationParts(assembly))
    {
        manager.ApplicationParts.Add(applicationPart);
    }
});

(which is similar code used by the [`AddApplicationPart`](https://github.com/aspnet/AspNetCore/blob/c7cb8467bfce721e2d66ef3862cd8c7c1fdbb421/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreMvcBuilderExtensions.cs#L100) method)

With this change, it only ensure the minimum requirements to run smidge without enabling everything that it doesn't necessarily need. 
```

Reply:
Looks good, I think. 

Does this mean that the call to addsmidge is now independent of position?
In my application, I'm calling :
```
services.AddSmidge()
  .AddControllersWithViews()
  .AddControllersAsServices() //for dependency injection as I'm using autofac
````
I remember there being issues with the smidge controllers not being found because AddControllersAsServices was called after AddSmidge, does this situation still work with the changes you made? I'm not able to debug this at the moment so can't tell for sure what the exact problem was... way to many balls in the air at the moment!